relay-general/src/protocol/measurements.rs

```diff
@@ -0,0 +1,199 @@
+use regex::Regex;
+
+use crate::types::{Annotated, Error, FromValue, Object, Value};
+
+lazy_static::lazy_static! {
+    static ref MEASUREMENT_NAME: Regex = Regex::new("^[a-z0-9-._]+$").unwrap();
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, ToValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Measurement {
+    /// Value of observed measurement value
+    #[metastructure(required = "true")]
+    pub value: Annotated<f64>,
+}
+
+// TODO: need bag_size?
+#[derive(Clone, Debug, Default, PartialEq, Empty, ToValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Measurements(pub Object<Measurement>);
+
+impl FromValue for Measurements {
+    fn from_value(value: Annotated<Value>) -> Annotated<Self> {
```

Comment:
You can simplify the implementation of this method considerably by delegating to the generic `FromValue` implementation of `Object`. Here's an outline of this (without adding the errors):

```rust
fn from_value(value: Annotated<Value>) -> Annotated<Self> {
  Object::<Measurement>::from_value(value).map_value(|measurements| {
    Self(measurements.retain(|name, _| is_valid_name(name)))
  });
}
```

Also, note that we prefer to construct `Annotated` via `Annotated::new()` as we're planning to change the tuple struct into a struct with fields.

Reply:
I've switched from the verbose match block to the form you suggested.